File: java/src15/org/nachodb/impl/AsyncReplicationMasterFile.java
package org.nachodb.impl;

import java.io.*;
import java.net.*;

import org.nachodb.*;

/**
 * File performing asynchronous replication of changed pages to specified slave nodes.
 */
public class AsyncReplicationMasterFile extends ReplicationMasterFile implements Runnable {
    /**
     * Constructor of replication master file
     * @param storage replication storage
     * @param file local file used to store data locally
     * @param asyncBufSize size of asynchronous buffer
     */
    public AsyncReplicationMasterFile(ReplicationMasterStorageImpl storage, IFile file, int asyncBufSize) { 
        super(storage, file);
        this.asyncBufSize = asyncBufSize;
        start();
    }


    /**
     * Constructor of replication master file
     * @param file local file used to store data locally
     * @param hosts slave node hosts to which replicastion will be performed
     * @param asyncBufSize size of asynchronous buffer
     * @param ack whether master should wait acknowledgment from slave node during trasanction commit
     */
    public AsyncReplicationMasterFile(IFile file, String[] hosts, int asyncBufSize, boolean ack) {                 
        super(file, hosts, ack);
        this.asyncBufSize = asyncBufSize;
        start();
    }

    private void start() {
        go = new Object();
        async = new Object();
        thread = new Thread(this);
        thread.start();
    }
                
    static class Parcel {
        byte[] data;
        long   pos;
        int    host;
        Parcel next;
    }
    
    public void write(long pos, byte[] buf) {
        file.write(pos, buf);
        for (int i = 0; i < out.length; i++) { 
            if (out[i] != null) {                
                byte[] data = new byte[8 + buf.length];
                Bytes.pack8(data, 0, pos);
                System.arraycopy(buf, 0, data, 8, buf.length);
                Parcel p = new Parcel();
                p.data = data;
                p.pos = pos;
                p.host = i;

                try { 
                    synchronized(async) { 
                        buffered += data.length;
                        while (buffered > asyncBufSize) { 
                            async.wait();
                        }
                    }
                } catch (InterruptedException x) {}
                    
                synchronized(go) { 
                    if (head == null) { 
                        head = tail = p;
                    } else { 
                        tail = tail.next = p;
                    }
                    go.notify();
                }
            }
        }
    }

    public void run() { 
        try { 
            while (true) { 
                Parcel p;
                synchronized(go) {
                    while (head == null) { 
                        if (closed) { 
                            return;
                        }
                        go.wait();
                    }
                    p = head;
                    head = p.next;
                }  
                
                synchronized(async) { 
                    if (buffered > asyncBufSize) { 
                        async.notifyAll();
                    }
                    buffered -= p.data.length;
                }

                int i = p.host;
                while (out[i] != null) { 
                    try { 
                        out[i].write(p.data);
                        if (!ack || p.pos != 0 || in[i].read(rcBuf) == 1) {
                            break;
                        }
                    } catch (IOException x) {}
                    
                    out[i] = null;
                    sockets[i] = null;
                    nHosts -= 1;
                    if (handleError(hosts[i])) { 
                        connect(i);
                    } else { 
                        break;
                    }
                }
            }
        } catch (InterruptedException x) {}
    }

    public void close() {
        try { 
            synchronized (go) {
                closed = true;
                go.notify();
            }
            thread.join();
        } catch (InterruptedException x) {}
        super.close();
    }

    private int     asyncBufSize;
    private int     buffered;
    private boolean closed;
    private Object  go;
    private Object  async;
    private Parcel  head;
    private Parcel  tail;    
    private Thread  thread;
}

File: csharp/src/impl/AsyncReplicationMasterFile.cs
#if WITH_REPLICATION
namespace Volante.Impl
{
    using System;
    using System.Threading;
    using System.Net;
    using System.Net.Sockets;
    using Volante;

    /// <summary>
    /// File performing asynchronous replication of changed pages to specified slave nodes.
    /// </summary>
    public class AsyncReplicationMasterFile : ReplicationMasterFile
    {
        /// <summary>
        /// Constructor of replication master file
        /// <param name="db">replication database</param>
        /// <param name="file">local file used to store data locally</param>
        /// <param name="asyncBufSize">size of asynchronous buffer</param>
        /// </summary>
        public AsyncReplicationMasterFile(ReplicationMasterDatabaseImpl db, IFile file, int asyncBufSize)
            : base(db, file)
        {
            this.asyncBufSize = asyncBufSize;
            start();
        }

        /// <summary>
        /// Constructor of replication master file
        /// <param name="file">local file used to store data locally</param>
        /// <param name="hosts">slave node hosts to which replication will be performed</param>
        /// <param name="asyncBufSize">size of asynchronous buffer</param>
        /// <param name="ack">whether master should wait acknowledgment from slave node during trasanction commit</param>
        /// </summary>
        public AsyncReplicationMasterFile(IFile file, String[] hosts, int asyncBufSize, bool ack)
            : base(file, hosts, ack)
        {
            this.asyncBufSize = asyncBufSize;
            start();
        }

        private void start()
        {
            go = new object();
            async = new object();
            thread = new Thread(new ThreadStart(run));
            thread.Start();
        }

        class Parcel
        {
            public byte[] data;
            public long pos;
            public int host;
            public Parcel next;
        }

        public override void Write(long pos, byte[] buf)
        {
            file.Write(pos, buf);
            for (int i = 0; i < sockets.Length; i++)
            {
                if (sockets[i] != null)
                {
                    byte[] data = new byte[8 + buf.Length];
                    Bytes.pack8(data, 0, pos);
                    Array.Copy(buf, 0, data, 8, buf.Length);
                    Parcel p = new Parcel();
                    p.data = data;
                    p.pos = pos;
                    p.host = i;

                    lock (async)
                    {
                        buffered += data.Length;
                        while (buffered > asyncBufSize)
                        {
                            Monitor.Wait(async);
                        }
                    }

                    lock (go)
                    {
                        if (head == null)
                        {
                            head = tail = p;
                        }
                        else
                        {
                            tail = tail.next = p;
                        }
                        Monitor.Pulse(go);
                    }
                }
            }
        }

        public void run()
        {
            while (true)
            {
                Parcel p;
                lock (go)
                {
                    while (head == null)
                    {
                        if (closed)
                        {
                            return;
                        }
                        Monitor.Wait(go);
                    }
                    p = head;
                    head = p.next;
                }

                lock (async)
                {
                    if (buffered > asyncBufSize)
                    {
                        Monitor.PulseAll(async);
                    }
                    buffered -= p.data.Length;
                }
                int i = p.host;
                while (sockets[i] != null)
                {
                    try
                    {
                        sockets[i].Send(p.data);
                        if (!ack || p.pos != 0 || sockets[i].Receive(rcBuf) == 1)
                        {
                            break;
                        }
                    }
                    catch (SocketException) { }

                    sockets[i] = null;
                    nHosts -= 1;
                    if (HandleError(hosts[i]))
                    {
                        connect(i);
                    }
                    else
                    {
                        break;
                    }
                }
            }
        }

        public override void Close()
        {
            lock (go)
            {
                closed = true;
                Monitor.Pulse(go);
            }
            thread.Join();
            base.Close();
        }

        private int asyncBufSize;
        private int buffered;
        private bool closed;
        private object go;
        private object async;
        private Parcel head;
        private Parcel tail;
        private Thread thread;
    }
}
#endif
